### ISSUE TITLE
Idea to improve load time for Validate, Gallery, and User Dashboard

### ISSUE THREAD

**Issue by misaugstad** [ProjectSidewalk/SidewalkWebpage#3004]
##### Brief description of problem/feature
One of the longest contributors to the page load time for Validate, Gallery, and the User Dashboard is the process that we go through to check that the images that we want to show are actually available from Google at that time. This greatly simplifies and speeds up our front-end, since we don't have to fail to show images and then continually ask for new ones to try.

However, I think that we can skip this step for some images in order to speed up the page load times. Right now we are doing this check for _every_ image. But I am proposing that we skip this check for images where we already verified that imagery exists very recently. This could even just be images that we checked within the past 24 hours!

We would see the biggest speedup on Validate and especially during mapathons. During mapathons, usually there is a lot of labeling _and_ a lot of validating of those labels that were just placed within the past hour. And I think that we can pretty safely assume that if we added a label to an image an hour ago on the Explore page, that we should be able to access that image again on Validate.

##### Potential solution(s)
We already record the last time that we successfully viewed the image in the `gsv_data` table, so it should be easy to just add a check for that timestamp before we start querying Google for the existence o the imagery.

**Comment by jonfroehlich** [comment]
FWIW, I've found that Gallery is actually quite performant. Was using it on our Seattle dataset the other day to evaluate "parked car" issues.

https://sidewalk-sea.cs.washington.edu/gallery?labelType=Obstacle&tags=parked%20car&validationOptions=correct

**Comment by misaugstad** [comment]
I measured this to figure out what the speedup would actually be. Short version: the ceiling is ~200-330 ms per page load (12-17%), but you only get most of it when an *entire* batch skips the check — and the backup-pano work since this issue was filed changes what a cache needs to store.

Methodology: temporarily instrumented `checkImageryBatch` and `gsvPanoExists`, ran Validate and the Gallery label endpoint against the local Chicago dev DB (a prod dump), then A/B'd with the check short-circuited entirely. Play dev mode, one machine's network path to Google, so treat absolute page times as indicative and the deltas as the real result.

### How many checks happen per page

`checkImageryBatch` fires them with `Future.traverse`, so a batch runs fully in parallel. That turns out to be the whole story.

| Page | Batch shape | API calls |
|---|---|---|
| Validate | mission of 10, query oversamples 5x | **50** |
| Gallery (`POST /label/labels`, n=30) | 7 label types in parallel, 20 each | **140** |
| User dashboard | same shape, per label type | n x 5 per type |

### What one check costs

- **Single metadata call:** p50 **86 ms**, p90 175 ms, p99 214 ms, max 428 ms (n=1870, fired in 50-way bursts). A warm single-connection curl is ~16-80 ms, so most of that p50 is our own concurrency, not Google being slow.
- **Per-check DB update** (`expired`/`has_backup`/`last_checked`): p50 3 ms, mean 4.6 ms — times 50 or 140 per request.
- **Whole check stage:** Validate p50 **188 ms** (145-295); Gallery 128-537 ms per type-batch.

### Maximum possible speedup

8 runs each, medians, with the imagery check short-circuited:

| | checks on | checks skipped | saving |
|---|---|---|---|
| `/validate` | 1.55 s | 1.33 s | **-0.21 s (13%)** |
| `POST /label/labels` (n=30) | 1.95 s | 1.71 s | **-0.24 s (12%)**; means -0.33 s (17%) |

### The catch: partial cache hits buy almost nothing

Since the calls are parallel, the stage costs *the max of k calls*, not the sum. Bootstrapping from the measured per-call latency distribution:

| cache hit rate | calls left | stage cost | saved |
|---|---|---|---|
| 0% | 50 | 213 ms | 0 |
| 34% | 33 | 200 ms | 13 ms |
| 50% | 25 | 191 ms | 22 ms |
| 90% | 5 | 142 ms | 71 ms |
| 98% | 1 | 88 ms | 126 ms |
| 100% | 0 | 0 ms | **214 ms** |

So we get essentially the whole win only when a whole batch skips — which is exactly the mapathon scenario described above, and close to nothing in steady-state mixed traffic.

Observed revisit rates, from validation history (fraction of validations where the same pano had another validation in the prior window). This is a **lower bound** — it ignores the 5x oversampled labels that get checked but never served, plus Gallery, dashboard, and Explore, all of which also refresh `last_checked`:

| | prior 1 h | prior 24 h | prior 7 d |
|---|---|---|---|
| Chicago (361k validations) | 28% | 34% | 42% |
| Seattle (439k validations) | 15% | 20% | 30% |

The premise holds mechanically: `PanoDataTable.updateFromExplore` writes `last_checked = lastViewed`, so labeling on Explore does prime the cache.

### How backup panos change the calculus

This is arguably bigger than the timing. `LabelTable.imageryViewable` now admits `expired = true` panos when a usable backup exists, so a large slice of the queue is panos **we already know are dead** — and we ask Google every single time just to be told so, then do the disk + DB lookup for the backup anyway:

- Chicago: **57,267 / 152,698 (37%)** of validatable labels sit on known-expired panos
- Seattle: **170,745 / 267,492 (64%)**

Consequences:

1. A cache keyed only on "recently verified as *good*" leaves 37-64% of each batch uncached, which per the table above is close to no speedup at all. **The cache has to cover the negative result too** — `pano_data` already stores `expired` alongside `last_checked`, so "checked within 24h" should short-circuit both ways, sending known-expired panos straight to the backup path. That's what pushes hit rates into the 90-100% band where the win actually shows up.
2. Those calls are 100% waste today regardless of latency.

### Read on whether it's worth doing

- Single-user latency: ~200 ms on Validate, ~250-330 ms on Gallery, and only with a near-fully-cached batch. Ordinary browsing gets maybe 10-20 ms.
- The real payoff is **under load**. 20 concurrent validators today means ~1,000 outbound HTTPS calls and ~1,000 `pano_data` UPDATEs in a burst against a 25-connection Hikari pool. During a mapathon the hit rate and the concurrency peak at the same time, so this is more a throughput/contention fix than a latency fix.
- It also removes tail risk: each call has a 5s timeout and we take that lottery 50 times per page. Max observed locally was 428 ms, but one stall on prod costs the whole page load.
- No dollar saving — Street View metadata requests are free.
- **If raw page time is the goal, there's a bigger lever:** the label query itself measured p50 **423 ms**, up to **1.66 s**, and Gallery fires 7 of them per load — 2-4x the cost of all the imagery checks combined.
